### ISSUE TITLE
OnDemand Validation.

### ISSUE THREAD

**Issue by babltiga** [palexdev/MaterialFX#102]
Is there a way to force on-demand validation? So say we have several fields which by default are marked as valid and a button that should trigger the validation and the state change ?

**Comment by Juke1999** [comment]
That is possible with a workaround: if you attach a listener to the regarding elements and the validator is being applied just before the validation check (on button press) and removed as soon as something changes in the elements. 
This is quite inconvenient, but it works. I'm using this for a login window.

**Comment by palexdev** [comment]
Sorry, but I don't quite understand the issue
Validators use bindings, they are automatically evaluated as soon as one of the conditions change or a condition/dependency is added, why would you force the update "programmatically"?

**Comment by palexdev** [comment]
@Juke1999 Ah yes, now I get it, thanks for the example
This is the intended behavior, as in many UI frameworks fields are validates as soon as they lose focus.

I wonder how I could improve it though... I mean, yes I could add a boolean flag to disable this behavior if needed, and you would be able to update the validator manually since they expose an update() method but... is this the right way to do it?

**Comment by Juke1999** [comment]
This is one way to do it, yes. But the "to evaluate" flag is perfectly fine, as there have to be different options on how to enable such validation features.
I will take a look myself in the following days and will comment as soon as I've found something promising.

**Comment by palexdev** [comment]
> This is one way to do it, yes. But the "to evaluate" flag is perfectly fine, as there have to be different options on how to enable such validation features.
> I will take a look myself in the following days and will comment as soon as I've found something promising.

Thank you very much 👍🏻

**Comment by palexdev** [comment]
@Juke1999 are you still interested in this?
I'm currently reviewing the validation API, and I was wondering if you would like to take a look at the new implementation

**Comment by Juke1999** [comment]
Sure, I can take a look.

**Comment by palexdev** [comment]
> Sure, I can take a look.

Thanks!
Here's a little demo of the new API: [validationdemo.zip](https://github.com/palexdev/MaterialFX/files/7794084/validationdemo.zip)

It's very experimental.
There are two validators. One is not automatic, meaning that it doesn't use JavaFX properties and it's also bound to one object type only. It also means that every time you validate an object it's needed to apply the specified predicate to the object which may be a little slower (nothing too noticeable though).
The other one is automatic, meaning that validations are made using JavaFX properties and expressions, so when you query the validator to check its state it is faster since the constraints updates automatically.

I kind of like the base idea of the "Constraint" objects, since it also allows to add many other details to the validation (such as the severity for example).
But to be honest the implementation still doesn't convince me. The automatic validator now is flexible enough to allow the user to decide when to trigger the validation. It can be set to be fully automatic (an example with explanation is given in the ValidationTest class), or can be triggered manually (also in the ValidationTest class, see the button)

I was kind of inspired by this awesome library [YAVI](https://github.com/making/yavi), but it's super simplified.

**Comment by Juke1999** [comment]
> > Sure, I can take a look.
> 
> Thanks!
> Here's a little demo of the new API: [validationdemo.zip](https://github.com/palexdev/MaterialFX/files/7794084/validationdemo.zip)
> 
> It's very experimental.
> There are two validators. One is not automatic, meaning that it doesn't use JavaFX properties and it's also bound to one object type only. It also means that every time you validate an object it's needed to apply the specified predicate to the object which may be a little slower (nothing too noticeable though).
> The other one is automatic, meaning that validations are made using JavaFX properties and expressions, so when you query the validator to check its state it is faster since the constraints updates automatically.
> 
> I kind of like the base idea of the "Constraint" objects, since it also allows to add many other details to the validation (such as the severity for example).
> But to be honest the implementation still doesn't convince me. The automatic validator now is flexible enough to allow the user to decide when to trigger the validation. It can be set to be fully automatic (an example with explanation is given in the ValidationTest class), or can be triggered manually (also in the ValidationTest class, see the button)
> 
> I was kind of inspired by this awesome library [YAVI](https://github.com/making/yavi), but it's super simplified.
> 
> Edit: Oh, I also added some experimental utils, don't know if they will remain the same but they seem to be quite useful

Alright.
I will take a look as soon as I'm back from vacation (09.01.) and will keep you updated about it.

**Comment by palexdev** [comment]
Hello @Juke1999!
I made some little changes to the automatic validator, and I think it could be the final version
Here's the same project, the new demo is ValidationTest2 and the new validator is in the validation2 package

Like I said just minor changes. The constraints are now kept in an observable list rather than a map, I think there's no need to associate constraints to a "name", also it's observable so that when it changes (add/remove constraints) the validator state is updated automatically.
A custom property to track the validator state is not needed anymore. Every constraint updates the validator by calling the update() method which at the end calls onUpdated() and this happens every time, regardless the validator state.
That method calls a BiConsumer which can be specified by the user that accepts a boolean and a List of constraints (so the validator state and the invalid constraints produced by the validate() method)
So basically there are two ways to query the validator. One is automatic and relies on the onUpdated() method, it is as easy as setting the consumer. The other one is just by calling the validate() method which like before returns a List of invalid constraints. Of course if the list is empty the validator state is valid (true) otherwise it's invalid (false).
I think the new validator is very convenient

**Comment by Juke1999** [comment]
> Hello @Juke1999!
> I made some little changes to the automatic validator, and I think it could be the final version
> Here's the same project, the new demo is ValidationTest2 and the new validator is in the validation2 package
> [validationdemo.zip](https://github.com/palexdev/MaterialFX/files/7833135/validationdemo.zip)
> 
> Like I said just minor changes. The constraints are now kept in an observable list rather than a map, I think there's no need to associate constraints to a "name", also it's observable so that when it changes (add/remove constraints) the validator state is updated automatically.
> A custom property to track the validator state is not needed anymore. Every constraint updates the validator by calling the update() method which at the end calls onUpdated() and this happens every time, regardless the validator state.
> That method calls a BiConsumer which can be specified by the user that accepts a boolean and a List of constraints (so the validator state and the invalid constraints produced by the validate() method)
> So basically there are two ways to query the validator. One is automatic and relies on the onUpdated() method, it is as easy as setting the consumer. The other one is just by calling the validate() method which like before returns a List of invalid constraints. Of course if the list is empty the validator state is valid (true) otherwise it's invalid (false).
> I think the new validator is very convenient
> 
> Edit: Oh I forgot to change one thing in the project. The valid property of the validator must be set to true initially (change the constructor) otherwise controls (such the stepper) querying an empty validator won't proceed since the return value is false

I'm home tomorrow and will have a look at the project. Will keep you updated!

**Comment by palexdev** [comment]
Yep no hurry, thanks!

I'm currently remaking the demo

**Comment by Juke1999** [comment]
Hey Alex,
I looked into the validators and they are looking good and are logical. I will write some test cases and will check if it there is anything missing or wrong with it.
Will keep you updated!

**Comment by Juke1999** [comment]
Hey @palexdev,
I looked into the validators and they are looking good and are logical. I will write some test cases and will check if it there is anything missing or wrong with it.
Will keep you updated!

**Comment by palexdev** [comment]
The new Validation API is not available on the staging branch
Will implement changes/suggestions if needed

**Comment by Juke1999** [comment]
Alright, it looks good and handles validations fine.
I will be there to help, if there are extensions required.

**Comment by palexdev** [comment]
The new Validation API is now available in the latest release, closing this issue
